File: backend/src/models/Dose.js
const mongoose = require('mongoose');

const doseSchema = new mongoose.Schema({
  doseId: {
    type: Number,
    unique: true
  },
  name: {
    type: String,
    required: false,
    trim: true
  },
  minAge: {
    type: Number,
    required: true,
    min: 0
  },
  maxAge: {
    type: Number,
    required: true,
    min: 0
  },
  minGap: {
    type: Number,
    required: true,
    min: 0,
    default: 0
  },
  vaccineID: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Vaccine',
    required: false // Changed to false to handle existing null values
  }
}, {
  timestamps: true
});

// Create indexes for better performance
doseSchema.index({ doseId: 1 });
doseSchema.index({ vaccineID: 1 });

// Auto-increment doseId before saving
doseSchema.pre('save', async function(next) {
  if (this.isNew && !this.doseId) {
    try {
      // Find the highest existing doseId and increment it
      const lastDose = await this.constructor.findOne({}, {}, { sort: { 'doseId': -1 } });
      this.doseId = lastDose ? lastDose.doseId + 1 : 1;
    } catch (error) {
      return next(error);
    }
  }
  next();
});

// Validate that vaccineID exists in Vaccine collection (only if provided)
doseSchema.pre('save', async function(next) {
  try {
    if (this.vaccineID) {
      const Vaccine = mongoose.model('Vaccine');
      const vaccine = await Vaccine.findById(this.vaccineID);
      if (!vaccine) {
        throw new Error('Vaccine with this ID does not exist');
      }
    }
    next();
  } catch (error) {
    next(error);
  }
});

module.exports = mongoose.model('Dose', doseSchema);

File: doctor/lib/models/dose.dart
class Dose {
  final String? id;
  final int? doseId;
  final String? name;
  final int minAge;
  final int maxAge;
  final int minGap;
  final String? vaccineID;
  final VaccineInfo? vaccine;

  Dose({
    this.id,
    this.doseId,
    this.name,
    required this.minAge,
    required this.maxAge,
    this.minGap = 0,
    this.vaccineID,
    this.vaccine,
  });

  factory Dose.fromJson(Map<String, dynamic> json) {
    return Dose(
      id: json['_id'] as String?,
      doseId: (json['doseId'] as num?)?.toInt(),
      name: json['name'] as String?,
      minAge: ((json['minAge'] as num?) ?? 0).toInt(),
      maxAge: ((json['maxAge'] as num?) ?? 0).toInt(),
      minGap: (json['minGap'] as num?)?.toInt() ?? 0,
      vaccineID: json['vaccineID'] == null
          ? null
          : json['vaccineID'] is String
              ? json['vaccineID'] as String?
              : (json['vaccineID'] as Map<String, dynamic>?)?['_id'] as String?,
      vaccine: json['vaccineID'] != null && json['vaccineID'] is Map
          ? VaccineInfo.fromJson(json['vaccineID'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        if (doseId != null) 'doseId': doseId,
        if (name != null) 'name': name,
        'minAge': minAge,
        'maxAge': maxAge,
        'minGap': minGap,
        if (vaccineID != null) 'vaccineID': vaccineID,
      };
}

class VaccineInfo {
  final String? id;
  final String? name;
  final int? vaccineID;

  VaccineInfo({
    this.id,
    this.name,
    this.vaccineID,
  });

  factory VaccineInfo.fromJson(Map<String, dynamic> json) {
    return VaccineInfo(
      id: json['_id'] as String?,
      name: json['name'] as String?,
      vaccineID: (json['vaccineID'] as num?)?.toInt(),
    );
  }
}
